File: kotlin/lib/src/main/kotlin/models/DestinationPatch.kt
// This file is @generated
package com.svix.kotlin.models

import com.svix.kotlin.MaybeUnset
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.buildJsonObject

@Serializable(with = DestinationPatchSerializer::class)
data class DestinationPatch(
    /** The Destination's UID. */
    val uid: MaybeUnset<String> = MaybeUnset.Unset,
    val status: MaybeUnset<DestinationStatusIn> = MaybeUnset.Unset,
    val batchSize: MaybeUnset<UShort> = MaybeUnset.Unset,
    val maxWaitSecs: MaybeUnset<UShort> = MaybeUnset.Unset,
    val eventTypes: List<String>? = null,
    val channels: List<String>? = null,
    val metadata: Map<String, String>? = null,
    val config: DestinationPatchConfig,
)

@Serializable
sealed class DestinationPatchConfig {
    val variantName: String
        get() = this::class.annotations.filterIsInstance<VariantName>().first().name

    abstract fun toJsonElement(): JsonElement

    @VariantName("pollingEndpoint")
    data object PollingEndpoint : DestinationPatchConfig() {
        override fun toJsonElement() = buildJsonObject {}
    }

    @VariantName("azureBlobStorage")
    data class AzureBlobStorage(val azureBlobStorage: AzureBlobStorageConfigPatch) :
        DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(AzureBlobStorageConfigPatch.serializer(), azureBlobStorage)
    }

    @VariantName("otelTracing")
    data class OtelTracing(val otelTracing: OtelTracingConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(OtelTracingConfigPatch.serializer(), otelTracing)
    }

    @VariantName("fifoEndpoint")
    data class FifoEndpoint(val fifoEndpoint: SinkHttpConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(SinkHttpConfigPatch.serializer(), fifoEndpoint)
    }

    @VariantName("amazonS3")
    data class AmazonS3(val amazonS3: S3ConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(S3ConfigPatch.serializer(), amazonS3)
    }

    @VariantName("googleCloudStorage")
    data class GoogleCloudStorage(val googleCloudStorage: GoogleCloudStorageConfigPatch) :
        DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(GoogleCloudStorageConfigPatch.serializer(), googleCloudStorage)
    }

    @VariantName("googleCloudPubSub")
    data class GoogleCloudPubSub(val googleCloudPubSub: GoogleCloudPubSubConfigPatch) :
        DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(GoogleCloudPubSubConfigPatch.serializer(), googleCloudPubSub)
    }

    @VariantName("sqs")
    data class Sqs(val sqs: SqsConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() = Json.encodeToJsonElement(SqsConfigPatch.serializer(), sqs)
    }

    @VariantName("sns")
    data class Sns(val sns: SnsConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() = Json.encodeToJsonElement(SnsConfigPatch.serializer(), sns)
    }

    @VariantName("bigQuery")
    data class BigQuery(val bigQuery: BigQueryConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(BigQueryConfigPatch.serializer(), bigQuery)
    }

    @VariantName("clickhouse")
    data class Clickhouse(val clickhouse: ClickhouseConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(ClickhouseConfigPatch.serializer(), clickhouse)
    }

    @VariantName("eventBridge")
    data class EventBridge(val eventBridge: EventBridgeConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(EventBridgeConfigPatch.serializer(), eventBridge)
    }

    @VariantName("snowflake")
    data class Snowflake(val snowflake: SnowflakeConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(SnowflakeConfigPatch.serializer(), snowflake)
    }

    @VariantName("rabbitMq")
    data class RabbitMq(val rabbitMq: RabbitMqConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(RabbitMqConfigPatch.serializer(), rabbitMq)
    }

    @VariantName("redshift")
    data class Redshift(val redshift: RedshiftConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(RedshiftConfigPatch.serializer(), redshift)
    }

    @VariantName("postgres")
    data class Postgres(val postgres: PostgresConfigPatch) : DestinationPatchConfig() {
        override fun toJsonElement() =
            Json.encodeToJsonElement(PostgresConfigPatch.serializer(), postgres)
    }

    companion object {
        private val typeMap =
            mapOf<String, (JsonElement) -> DestinationPatchConfig>(
                "pollingEndpoint" to { _ -> PollingEndpoint },
                "azureBlobStorage" to
                    { config ->
                        AzureBlobStorage(
                            Json.decodeFromJsonElement(
                                AzureBlobStorageConfigPatch.serializer(),
                                config,
                            )
                        )
                    },
                "otelTracing" to
                    { config ->
                        OtelTracing(
                            Json.decodeFromJsonElement(OtelTracingConfigPatch.serializer(), config)
                        )
                    },
                "fifoEndpoint" to
                    { config ->
                        FifoEndpoint(
                            Json.decodeFromJsonElement(SinkHttpConfigPatch.serializer(), config)
                        )
                    },
                "amazonS3" to
                    { config ->
                        AmazonS3(Json.decodeFromJsonElement(S3ConfigPatch.serializer(), config))
                    },
                "googleCloudStorage" to
                    { config ->
                        GoogleCloudStorage(
                            Json.decodeFromJsonElement(
                                GoogleCloudStorageConfigPatch.serializer(),
                                config,
                            )
                        )
                    },
                "googleCloudPubSub" to
                    { config ->
                        GoogleCloudPubSub(
                            Json.decodeFromJsonElement(
                                GoogleCloudPubSubConfigPatch.serializer(),
                                config,
                            )
                        )
                    },
                "sqs" to
                    { config ->
                        Sqs(Json.decodeFromJsonElement(SqsConfigPatch.serializer(), config))
                    },
                "sns" to
                    { config ->
                        Sns(Json.decodeFromJsonElement(SnsConfigPatch.serializer(), config))
                    },
                "bigQuery" to
                    { config ->
                        BigQuery(
                            Json.decodeFromJsonElement(BigQueryConfigPatch.serializer(), config)
                        )
                    },
                "clickhouse" to
                    { config ->
                        Clickhouse(
                            Json.decodeFromJsonElement(ClickhouseConfigPatch.serializer(), config)
                        )
                    },
                "eventBridge" to
                    { config ->
                        EventBridge(
                            Json.decodeFromJsonElement(EventBridgeConfigPatch.serializer(), config)
                        )
                    },
                "snowflake" to
                    { config ->
                        Snowflake(
                            Json.decodeFromJsonElement(SnowflakeConfigPatch.serializer(), config)
                        )
                    },
                "rabbitMq" to
                    { config ->
                        RabbitMq(
                            Json.decodeFromJsonElement(RabbitMqConfigPatch.serializer(), config)
                        )
                    },
                "redshift" to
                    { config ->
                        Redshift(
                            Json.decodeFromJsonElement(RedshiftConfigPatch.serializer(), config)
                        )
                    },
                "postgres" to
                    { config ->
                        Postgres(
                            Json.decodeFromJsonElement(PostgresConfigPatch.serializer(), config)
                        )
                    },
            )

        fun fromTypeAndConfig(type: String, config: JsonElement): DestinationPatchConfig {
            return typeMap[type]?.invoke(config)
                ?: throw SerializationException("Unknown type: $type")
        }
    }
}

class DestinationPatchSerializer : KSerializer<DestinationPatch> {
    @Serializable
    private data class DestinationPatchSurrogate(
        /** The Destination's UID. */
        val uid: MaybeUnset<String> = MaybeUnset.Unset,
        val status: MaybeUnset<DestinationStatusIn> = MaybeUnset.Unset,
        val batchSize: MaybeUnset<UShort> = MaybeUnset.Unset,
        val maxWaitSecs: MaybeUnset<UShort> = MaybeUnset.Unset,
        val eventTypes: List<String>? = null,
        val channels: List<String>? = null,
        val metadata: Map<String, String>? = null,
        val type: String,
        val config: JsonElement = buildJsonObject {},
    )

    override val descriptor: SerialDescriptor = DestinationPatchSurrogate.serializer().descriptor

    override fun serialize(encoder: Encoder, value: DestinationPatch) {
        val surrogate =
            DestinationPatchSurrogate(
                uid = value.uid,
                status = value.status,
                batchSize = value.batchSize,
                maxWaitSecs = value.maxWaitSecs,
                eventTypes = value.eventTypes,
                channels = value.channels,
                metadata = value.metadata,
                type = value.config.variantName,
                config = value.config.toJsonElement(),
            )
        encoder.encodeSerializableValue(DestinationPatchSurrogate.serializer(), surrogate)
    }

    override fun deserialize(decoder: Decoder): DestinationPatch {
        val surrogate = decoder.decodeSerializableValue(DestinationPatchSurrogate.serializer())
        return DestinationPatch(
            uid = surrogate.uid,
            status = surrogate.status,
            batchSize = surrogate.batchSize,
            maxWaitSecs = surrogate.maxWaitSecs,
            eventTypes = surrogate.eventTypes,
            channels = surrogate.channels,
            metadata = surrogate.metadata,
            config = DestinationPatchConfig.fromTypeAndConfig(surrogate.type, surrogate.config),
        )
    }
}

File: javascript/src/models/destinationPatch.ts
// this file is @generated
import {
  type AzureBlobStorageConfigPatch,
  AzureBlobStorageConfigPatchSerializer,
} from "./azureBlobStorageConfigPatch";
import {
  type BigQueryConfigPatch,
  BigQueryConfigPatchSerializer,
} from "./bigQueryConfigPatch";
import {
  type ClickhouseConfigPatch,
  ClickhouseConfigPatchSerializer,
} from "./clickhouseConfigPatch";
import {
  type DestinationStatusIn,
  DestinationStatusInSerializer,
} from "./destinationStatusIn";
import {
  type EventBridgeConfigPatch,
  EventBridgeConfigPatchSerializer,
} from "./eventBridgeConfigPatch";
import {
  type GoogleCloudPubSubConfigPatch,
  GoogleCloudPubSubConfigPatchSerializer,
} from "./googleCloudPubSubConfigPatch";
import {
  type GoogleCloudStorageConfigPatch,
  GoogleCloudStorageConfigPatchSerializer,
} from "./googleCloudStorageConfigPatch";
import {
  type OtelTracingConfigPatch,
  OtelTracingConfigPatchSerializer,
} from "./otelTracingConfigPatch";
import {
  type PostgresConfigPatch,
  PostgresConfigPatchSerializer,
} from "./postgresConfigPatch";
import {
  type RabbitMqConfigPatch,
  RabbitMqConfigPatchSerializer,
} from "./rabbitMqConfigPatch";
import {
  type RedshiftConfigPatch,
  RedshiftConfigPatchSerializer,
} from "./redshiftConfigPatch";
import { type S3ConfigPatch, S3ConfigPatchSerializer } from "./s3ConfigPatch";
import {
  type SinkHttpConfigPatch,
  SinkHttpConfigPatchSerializer,
} from "./sinkHttpConfigPatch";
import {
  type SnowflakeConfigPatch,
  SnowflakeConfigPatchSerializer,
} from "./snowflakeConfigPatch";
import { type SnsConfigPatch, SnsConfigPatchSerializer } from "./snsConfigPatch";
import { type SqsConfigPatch, SqsConfigPatchSerializer } from "./sqsConfigPatch";

interface _DestinationPatchFields {
  /** The Destination's UID. */
  uid?: string | null;
  status?: DestinationStatusIn | null;
  batchSize?: number | null;
  maxWaitSecs?: number | null;
  eventTypes?: string[];
  channels?: string[];
  metadata?: { [key: string]: string };
}

// biome-ignore lint/suspicious/noEmptyInterface: backwards compat
interface DestinationPatchPollingEndpointConfig {}

interface DestinationPatchPollingEndpoint {
  type: "pollingEndpoint";
  config?: DestinationPatchPollingEndpointConfig;
}

interface DestinationPatchAzureBlobStorage {
  type: "azureBlobStorage";
  config: AzureBlobStorageConfigPatch;
}

interface DestinationPatchOtelTracing {
  type: "otelTracing";
  config: OtelTracingConfigPatch;
}

interface DestinationPatchFifoEndpoint {
  type: "fifoEndpoint";
  config: SinkHttpConfigPatch;
}

interface DestinationPatchAmazonS3 {
  type: "amazonS3";
  config: S3ConfigPatch;
}

interface DestinationPatchGoogleCloudStorage {
  type: "googleCloudStorage";
  config: GoogleCloudStorageConfigPatch;
}

interface DestinationPatchGoogleCloudPubSub {
  type: "googleCloudPubSub";
  config: GoogleCloudPubSubConfigPatch;
}

interface DestinationPatchSqs {
  type: "sqs";
  config: SqsConfigPatch;
}

interface DestinationPatchSns {
  type: "sns";
  config: SnsConfigPatch;
}

interface DestinationPatchBigQuery {
  type: "bigQuery";
  config: BigQueryConfigPatch;
}

interface DestinationPatchClickhouse {
  type: "clickhouse";
  config: ClickhouseConfigPatch;
}

interface DestinationPatchEventBridge {
  type: "eventBridge";
  config: EventBridgeConfigPatch;
}

interface DestinationPatchSnowflake {
  type: "snowflake";
  config: SnowflakeConfigPatch;
}

interface DestinationPatchRabbitMq {
  type: "rabbitMq";
  config: RabbitMqConfigPatch;
}

interface DestinationPatchRedshift {
  type: "redshift";
  config: RedshiftConfigPatch;
}

interface DestinationPatchPostgres {
  type: "postgres";
  config: PostgresConfigPatch;
}

export type DestinationPatch = _DestinationPatchFields &
  (
    | DestinationPatchPollingEndpoint
    | DestinationPatchAzureBlobStorage
    | DestinationPatchOtelTracing
    | DestinationPatchFifoEndpoint
    | DestinationPatchAmazonS3
    | DestinationPatchGoogleCloudStorage
    | DestinationPatchGoogleCloudPubSub
    | DestinationPatchSqs
    | DestinationPatchSns
    | DestinationPatchBigQuery
    | DestinationPatchClickhouse
    | DestinationPatchEventBridge
    | DestinationPatchSnowflake
    | DestinationPatchRabbitMq
    | DestinationPatchRedshift
    | DestinationPatchPostgres
  );

export const DestinationPatchSerializer = {
  _fromJsonObject(object: any): DestinationPatch {
    const type = object["type"];

    function getConfig(type: string): any {
      switch (type) {
        case "pollingEndpoint":
          return {};
        case "azureBlobStorage":
          return AzureBlobStorageConfigPatchSerializer._fromJsonObject(object["config"]);
        case "otelTracing":
          return OtelTracingConfigPatchSerializer._fromJsonObject(object["config"]);
        case "fifoEndpoint":
          return SinkHttpConfigPatchSerializer._fromJsonObject(object["config"]);
        case "amazonS3":
          return S3ConfigPatchSerializer._fromJsonObject(object["config"]);
        case "googleCloudStorage":
          return GoogleCloudStorageConfigPatchSerializer._fromJsonObject(
            object["config"]
          );
        case "googleCloudPubSub":
          return GoogleCloudPubSubConfigPatchSerializer._fromJsonObject(object["config"]);
        case "sqs":
          return SqsConfigPatchSerializer._fromJsonObject(object["config"]);
        case "sns":
          return SnsConfigPatchSerializer._fromJsonObject(object["config"]);
        case "bigQuery":
          return BigQueryConfigPatchSerializer._fromJsonObject(object["config"]);
        case "clickhouse":
          return ClickhouseConfigPatchSerializer._fromJsonObject(object["config"]);
        case "eventBridge":
          return EventBridgeConfigPatchSerializer._fromJsonObject(object["config"]);
        case "snowflake":
          return SnowflakeConfigPatchSerializer._fromJsonObject(object["config"]);
        case "rabbitMq":
          return RabbitMqConfigPatchSerializer._fromJsonObject(object["config"]);
        case "redshift":
          return RedshiftConfigPatchSerializer._fromJsonObject(object["config"]);
        case "postgres":
          return PostgresConfigPatchSerializer._fromJsonObject(object["config"]);
        default:
          throw new Error(`Unexpected type: ${type}`);
      }
    }

    return {
      type,
      config: getConfig(type),
      uid: object["uid"],
      status:
        object["status"] != null
          ? DestinationStatusInSerializer._fromJsonObject(object["status"])
          : undefined,
      batchSize: object["batchSize"],
      maxWaitSecs: object["maxWaitSecs"],
      eventTypes: object["eventTypes"],
      channels: object["channels"],
      metadata: object["metadata"],
    };
  },

  _toJsonObject(self: DestinationPatch): any {
    // biome-ignore lint/suspicious/noImplicitAnyLet: the return type needs to be any
    let config;
    switch (self.type) {
      case "pollingEndpoint":
        config = {};
        break;
      case "azureBlobStorage":
        config = AzureBlobStorageConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "otelTracing":
        config = OtelTracingConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "fifoEndpoint":
        config = SinkHttpConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "amazonS3":
        config = S3ConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "googleCloudStorage":
        config = GoogleCloudStorageConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "googleCloudPubSub":
        config = GoogleCloudPubSubConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "sqs":
        config = SqsConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "sns":
        config = SnsConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "bigQuery":
        config = BigQueryConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "clickhouse":
        config = ClickhouseConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "eventBridge":
        config = EventBridgeConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "snowflake":
        config = SnowflakeConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "rabbitMq":
        config = RabbitMqConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "redshift":
        config = RedshiftConfigPatchSerializer._toJsonObject(self.config);
        break;
      case "postgres":
        config = PostgresConfigPatchSerializer._toJsonObject(self.config);
        break;
    }

    return {
      type: self.type,
      config: config,
      uid: self.uid,
      status:
        self.status != null
          ? DestinationStatusInSerializer._toJsonObject(self.status)
          : undefined,
      batchSize: self.batchSize,
      maxWaitSecs: self.maxWaitSecs,
      eventTypes: self.eventTypes,
      channels: self.channels,
      metadata: self.metadata,
    };
  },
};
